src/main/webapp/resources/js/pages/projects/samples-metadata-import/services/importReducer.ts

```diff
@@ -0,0 +1,246 @@
+import {
+  createAction,
+  createAsyncThunk,
+  createReducer,
+} from "@reduxjs/toolkit";
+import { validateSampleName } from "../../../../apis/metadata/sample-utils";
+import {
+  createSample,
+  FieldUpdate,
+  MetadataItem,
+  updateSample,
+  ValidateSampleNameModel,
+  validateSamples,
+  ValidateSamplesResponse,
+} from "../../../../apis/projects/samples";
+import { ImportDispatch, ImportState } from "../store";
+
+interface MetadataValidateDetailsItem {
+  isSampleNameValid: boolean;
+  foundSampleId?: number;
+}
+
+interface MetadataSaveDetailsItem {
+  saved: boolean;
+  error?: string;
+}
+interface SaveMetadataResponse {
+  metadataSaveDetails: Record<string, MetadataSaveDetailsItem>;
+}
+
+interface SetSampleNameColumnResponse {
+  sampleNameColumn: string;
+  metadataValidateDetails: Record<string, MetadataValidateDetailsItem>;
+}
+
+export interface InitialState {
+  sampleNameColumn: string;
+  headers: string[];
+  metadata: MetadataItem[];
+  metadataValidateDetails: Record<string, MetadataValidateDetailsItem>;
+  metadataSaveDetails: Record<string, MetadataSaveDetailsItem>;
+}
+
+const initialState: InitialState = {
+  sampleNameColumn: "",
+  headers: [],
+  metadata: [],
+  metadataValidateDetails: {},
+  metadataSaveDetails: {},
+};
+
+/*
+Redux async thunk for saving the metadata to samples.
+For more information on redux async thunks see: https://redux-toolkit.js.org/api/createAsyncThunk
+ */
+export const saveMetadata = createAsyncThunk<
+  SaveMetadataResponse,
+  { projectId: string; selectedMetadataKeys: string[] },
+  { dispatch: ImportDispatch; state: ImportState }
+>(
+  `importReducer/saveMetadata`,
+  async ({ projectId, selectedMetadataKeys }, { dispatch, getState }) => {
+    const state: ImportState = getState();
+    const { sampleNameColumn, headers, metadata, metadataValidateDetails } =
+      state.importReducer;
+    const metadataSaveDetails: Record<string, MetadataSaveDetailsItem> = {};
+
+    const chunkSize = 100;
+    for (let i = 0; i < metadata.length; i = i + chunkSize) {
```

Comment:
This isn't actually batching.   What you will need to do is create a new endpoint on the server that can accept more than 1 sample with metadata.  Then, for example, if there are 1000 samples to update, send them in sets of 100.  If there are less than 100, then just send them all.  When the requests kick off display a progress bar to the user, after each batch is complete update the progress bar.

Reply:
The promise chaining did speed up the upload by ~6x. The progress bar is being updated with every chunk. I had very specific instructions from https://github.com/phac-nml/irida/issues/1350. I also thought your preference was to keep the processing in the front end. If I'm suppose to update the api endpoints to accept more than one request, can we please have a quick conversation? For example, I was asked to pass the sampleId as a path variable. I'm also sending the sample Id back after creation. Am I rolling back the whole request on a failure? Etc.

Reply:
Your speed-up is coming from the fact that previously you had an `await` statement for each sample upload (causing each to wait until the response comes back), now using a promise array you are actually parallelizing them. You are not changing any of the parsing of the file or front-end processing, just the upload.  We can leave it this way and updated the batching in a future PR.

Reply:
Sure, I'll wait for the PR that describes the specifics for the endpoint like #1350.